src/ccache.rs

```diff
@@ -227,6 +243,205 @@ impl TryFrom<&SessionKey> for KeyBlockV4 {
+                        EncryptedData::Aes256CtsHmacSha196 { kvno: _, data } => {
+                            let t = Asn1Ticket {
+                                tkt_vno: 5,
+                                realm: (&enc_part.server).try_into()?,
+                                sname: (&enc_part.server).try_into()?,
+                                enc_part: Asn1EncryptedData {
+                                    etype: Asn1EncryptionType::AES256_CTS_HMAC_SHA1_96 as i32,
+                                    kvno: Some(1), // TODO Why?
+                                    cipher: OctetString::new(data.clone())
+                                        .expect("Failed to build OctetString"),
+                                },
+                            };
+                            let tt = Asn1TaggedTicket::new(t);
+                            tt.to_der().map_err(|e| {
+                                println!("{:#?}", e);
+                                KrbError::DerEncodeEncTicketPart
+                            })?
+                        }
+                    },
+                },
+                second_ticket: DataComponent { value: vec![] },
+            }),
+        ];
+
+        let ccache: FileCredentialCacheV4 = FileCredentialCacheV4 {
+            header,
+            principal,
+            credentials,
+        };
+
+        Ok(FileCredentialCache::V4(ccache))
+    }
+
+    #[cfg(feature = "msextensions")]
+    pub fn from_asrep(krb_kdc_rep: KrbKdcRep, client_key: &DerivedKey) -> Result<Self, KrbError> {
+        let (name, ticket, kdc_reply): (Name, Ticket, KdcReplyPart) =
+            match KerberosReply::try_from(krb_kdc_rep).map_err(|e| {
+                println!("{:#?}", e);
+                KrbError::InsufficientData
+            })? {
+                KerberosReply::AS(AuthenticationReply {
+                    name,
+                    enc_part,
+                    pa_data,
+                    ticket,
+                }) => {
+                    let etype_info = pa_data
+                        .as_ref()
+                        .map(|pa_inner| pa_inner.etype_info2.as_slice());
+
+                    let kdc_reply = enc_part.decrypt_enc_kdc_rep(&client_key)?;
+                    (name, ticket, kdc_reply)
+                }
+                _ => unreachable!(),
+            };
+
+        FileCredentialCache::new(&name, &ticket, &kdc_reply, None)
+    }
+
+    pub fn store(self: Self, path: Option<PathBuf>) -> Result<(), KrbError> {
+        let mut f: File = match path {
+            Some(p) => File::create(p).map_err(|e| KrbError::IoError(e)),
+            None => match env::var("KRB5CCNAME") {
+                Ok(val) => {
+                    if !val.starts_with("FILE:") {
+                        return Err(KrbError::UnsupportedCredentialCacheType);
+                    }
+                    let p = val.strip_prefix("FILE:").expect("Failed to strip prefix");
```

Comment:
FYI, I'm thinking we should support DIR here, so we can stash the ccache in a file named `username@realm`, for example, within the DIR sub-directory.

Reply:
I also think we should be able to specify the DIR from the function call. We could have the `path` parameter be a enum:
```rust
enum CCachePath {
    FILE(String),
    DIR(String),
}
```
etc

Reply:
Probably better that we pursue the kernel keyring rather than directory here I think. The file mode is a good easy test, but keyring is likely more secure.

Reply:
It also occurs to me that this is better supported outside this library. Recall that the idprovider will be fetching the AsRep, but it's the task_daemon that will need to install the ccache somewhere, IIUC. So perhaps this library should just respond with a ccache as a `Vec<u8>`?

Reply:
I think that's not a good option - it would be better to respond with the ccache as a strong type because then the caller can read the file, get the cache, and then if they get session keys they can add them, remove them, manage them, inspect them etc.

If we made this `vec<u8>` we lose all that.